### DISCUSSION TITLE
Enabling self-pickup flow in ONDC

### DISCUSSION THREAD

**Discussion by BLR-0118** [ONDC-Official/protocol-network-extension#39] [General]
Some initial thoughts on enabling self-pickup flow in ONDC:
1. Catalog publishes list of items that are for delivery, pickup, pickup & delivery;
2. If buyer creates a cart with items that are for pickup / pickup & delivery, they can have the order marked for pickup;
3. Buyer app shows option for pickup / delivery (as appropriate);
4. Buyer selects pickup option;
5. Order will be processed & made available for buyer to pickup.

Open issues to be sorted out:
1. How does the merchant inform the buyer that order is ready for pickup? (we can reuse the ready_to_ship tag currently used with on-network logistics providers);
2. How does the merchant identify the buyer? (one option is OTP but need to communicate to the buyer app as well)
3. How does the buyer identify the order to be picked up? (pickup info details need to be sent to buyer app similar to what's being sent to logistics providers);
4. Order closure - who marks the order state as completed & how to ensure the state is updated only when the order is picked up?
5. How will the settlement flow work for cash on delivery?

**Comment by BLR-0118** [comment]
@aakasjayaraj @abaweja @batramanuj @mahoriR @mherle @VidyaGF @NajeebMohammed @ravi-kumar-yadav @ONDC-Official/participants

**Reply by NajeebMohammed** [reply]
@BLR-0118  some initial thoughts - why would an item be "pickup & delivery" ? Can you give an example? I thought an item could be marked either for pickup or delivery and the order will be marked as "pickup & delivery" if it contains both pickup and delivery items?

Also, if an order has both pickup and delivery items, how do we identify which items are for pickup and which ones are for delivery? The select API needs to be enhanced to cater for this.

**Reply by ONDC-Tech** [reply]
@NajeebMohammed - some items could be either pickup or delivery and the buyer gets to choose which option to use. Agree, some changes are required but please do advise on the open questions

**Comment by mherle** [comment]
What is the pickup flow @BLR-0118 ? Is this like picking up from customer/buyer (such as laundry) ?

**Reply by ONDC-Tech** [reply]
@mherle - yes, where the buyer picks up the items from the store after placing an order on ONDC

**Reply by mherle** [reply]
oh! - This is more like self-pickup by buyer. I'd suggest we not use the pickup word as this might be confusing for first-mile pickup. Should we call it _self-pickup_ ?

**Comment by aanay1** [comment]
Hi Folks, 
I am receiving notifications from GitHub about the updates. 
Am I required to participate, or is it by mistake?

Thought - We may want to consider - 
a. # of pickup stations and their capacity
b. inventory update
c. buyer's location and Tier of the city

**Comment by abaweja** [comment]
Suggestions on flow:

1. Catalog should have "pickup available", "delivery available". For pickup and delivery items both should be available. Items like sizzling brownie or a sizzler could be pickup available only, for items being delivered from cloud kitchen, it would be delivery available. for most of items and stores, both flags will be enabled.
2. A single order can only be pickup or delivery. If user mixes items, appropriate. warning should be shown. There is scope of confusion here, but pareto is in favour of all items in store being either delivery only (cloud kitchen) or both options available. There will be limited number of items which will be pickup only when store supports both pickup and delivery.

Agree with Mahesh, to avoid confusion for logistics terminology, Pickup can be "Self-Pickup"

**Comment by abaweja** [comment]
A pickup order will lead to creation of Fulfilment with type "Self-pickup". rest of flow remains identical to regular shipping. 
Whether OTP flow needs to be enforced should be available as option with NPs. In this case, buyer app can require OTP verification for delivery. OTP verification should be with the participant requesting OTP flow. So if buyer app mandated OTP verification on delivery, OTP needs to submitted back to BAP for verification and OTP verified gives closure to order.
This should be implemented not just for pickup flow but as generic construct in fulfilment for any order.
For COD orders, assuming continued operations and order flow, seller platform will deduct fees from future pre-paid order settlement and settle to BAP as well.

**Comment by BLR-0118** [comment]
@dashbitla-ib - for your comments

**Reply by dashbitla-ib** [reply]
Can we add Curbside also along with Pickup and Delivery. Curbside means, Seller will drop it to the Buyer directly in his vehicle. For Curbside - take the Buyer vehicle details as well.

Enable Support for Pickup Timeslots and Delivery Timeslots.

How does the merchant inform the buyer that order is ready for pickup? (we can reuse the ready_to_ship tag currently used with on-network logistics providers);
=> Yes, this is fine. If renaming is possible, change this to READY state, that will cover for both pickup and delivery.

How does the merchant identify the buyer? (one option is OTP but need to communicate to the buyer app as well)
=> Order Number, Mobile Number and Name all are there right? Do we really need another OTP? As a check Seller can check the last 5 digits of the mobile number for extra confirmation.

How does the buyer identify the order to be picked up? (pickup info details need to be sent to buyer app similar to what's being sent to logistics providers);
=> When the Order is ready, Buyer need to be notified.
=> Also, when the Buyer reached the Seller place, give an option to say "I am here" with live location shared. This can be used for both Pickup and Curbside orders to notify the retailer that Buyer is in their store.

Order closure - who marks the order state as completed & how to ensure the state is updated only when the order is picked up?
=> Seller can mark the Order as Pickup up once the Pickup is done. This is similar to Delivery Order, kind of trust based.

If we want to make it fool proof, generate an OTP code and share it to the Buyer. This is not visible to the Seller or LP. At the time of Pickup or Delivery, Seller or LP must verify this OTP to mark the Order as Completed. Amazon does this for certain high value Order Items.

How will the settlement flow work for cash on delivery?
=> this should be similar to any other CoD flow right? Instead of LP collecting he money, it goes to the Seller direct. 
Considering this - the settlement flow to be defined to settle for the Buyer Finder Fee and applicable Seller Platform Fee.

**Comment by aakasjayaraj** [comment]
Here are some fields that might be needed for pickup (Pickup instructions). 

Location of Pickup- Lat, long, complete address of the seller
Seller's contact number
Date & Time of pickup (Can be slots) or store timings
Pickup Code
Optional OTP - might be useful for high-value items
**Register/counter number**
**Vehicle info- for Curbside pickup**
**Instructions - For the buyer to specify special instructions**

Closing of the order, RTO (partial, full) can be similar to sellers doing their own delivery.

Also, shall the buyer have the option to signal to the seller that "they've arrived at the store" via the buyer app- since there's a chance that sellers might only start preparing orders at the last minute, this might help in improving the experience and also decrease the chance of mishaps & post delivery disputes - Any thoughts on this?

**Reply by abaweja** [reply]
All these constructs already available in Fulfilment object. There is validation construct which could be OTP and is optional. Essentially we should look at "Self pickup" as good as current Fulfilment, just that pickup and delivery gets short circuited and Fulfilment partner is same as customer. If there is anything missing in this construct, idea should be to add that in generic Fulfilment object itself for all kind of Fulfilment. So vehicle info for customer could be added in same way as a vehicle info for delivery partner etc.

**Comment by BLR-0118** [comment]
@bansalmeghaa - for your comments

**Reply by bansalmeghaa** [reply]
How does the merchant inform the buyer that the order is ready for pickup? (we can reuse the ready_to_ship tag currently used with on-network logistics providers);
 
_First, we should clearly identify the difference between pickup and self pickup. This one should be named as self pickup. The tag Ready_to_Ship is the right one to inform the buyer about the item ready for  pickup._
 
How does the merchant identify the buyer? (one option is OTP but it is need to communicate to the buyer app as well).
 
_The merchant should be able to identify the buyer with the order no,( which is unique)     and also can be with OTP  plus Order no in case of high value items.
            (This required order no will be communicated to the seller app from buyer app)_
 
 
How does the buyer identify the order to be picked up? (pickup info details need to be sent to buyer app similar to what's being sent to logistics providers);
 
_Buyer identify the order to be picked up at seller end through Network order no again and will be notified on buyer app about the address from where to be picked up ( self pickup here is equivalent to logistic player picking up)_
 
Order closure - who marks the order state as completed & how to ensure the state is updated only when the order is picked up?
 
_Order Closure can happen only when the seller enters the OTP in the app. The OTP comes to the buyer through the buyer app. That's how the seller will close the order state as completed and we ensure that it is updated only after being picked up._
 
How will the settlement flow work for cash on delivery? 
_For COD, instead of LP, Seller is collecting the money. Not Sure on the flow side._
 
Overall flow looks like:
 

- Buyer orders on buyer app  >>
-  choose options as self pickup ( from Self pickup or delivery option) >> 
- Buyer be notified on buyer app with ready_to_ ship tag that item is ready to be picked.>> 
- buyer also be provided with address info of seller>>
-  buyer goes to seller>
-  identifies itself as the right buyer through order no> 
- picks up the shipment >
-  seller enters the  OTP code that comes on buyer mobile no while picking up and seller marks it as picked up after OTP entering. OTP from the buyer app to the buyer can be triggered when the seller marks the shipment with the flag ready_to _ ship.

**Comment by BLR-0118** [comment]
1. For self-pickup, the buyer should also be informed of store timings & self-pickup slot (?)
2. Also, how exactly will the order be marked as completed? If seller marks it, how do they get the OTP, from whom and who verifies & marks it as completed?
@aakasjayaraj @bansalmeghaa @abaweja

**Reply by bansalmeghaa** [reply]
1. For self-pickup, the buyer should also be informed of store timings & self-pickup slot (?)

> _While choosing self pickup option for delivery, buyer should be shown the slots available to pickup and then the address of the store as well as timings of the store. This will be useful to avoid unsuccessful attempts by the buyer for pickup. Also, for the store to plan the no of pickup possible at their end.

2. Also, how exactly will the order be marked as completed? If seller marks it, how do they get the OTP, from whom and who verifies & marks it as completed?

>_ The order is valuable for seller and buyer. Picked up can be confirmed by buyer only, cz if  buyer does not pick and seller incorrectly marked picked up, it can create frauds. so it could be two step thing.
            a. Buyer confirms pickup and only after step this confirmation- 
            b. Seller confirms the pickup through OTP or by any other method. 
Seller cannot confirm before buyer confirmation. Buyer can confirm the pickup on buyer app and seller also confirms on his end on the platform where they received the order. Such transactional contract may be required till the whole  order cycle is completed.
